### PR TITLE
Implement clipped text directive

### PR DESCRIPTION
**Pull request by juanmendes**
Changed from a component into a directive to avoid creating unnecessary DOM nodes
and to make its API simpler. It also makes it easier to dynamically create it since
the component version required transcluded nodes that are difficult to achieve when
rendering components dynamically as we do on our grid.

clipped-text-directive

## Manual Testing
* Tooltip hides after a timeout when moving mouse away from host
  * Tooltip does not hide if you move the mouse quickly to the tooltip
* Tooltip displays in a direction that attempts to keep the tooltip on screen, that is, it is towards the opposite screen quadrant.

## Coverage
100% Statements 98/98
96.15% Branches 25/26
100% Functions 13/13
100% Lines 97/97

Signed-off-by: Juan Mendes <jmendes@vmware.com>

### CODE REVIEW

**Comment by jakedsouza** [comment]
# [Codecov](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/42?src=pr&el=h1) Report
> Merging [#42](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/42?src=pr&el=desc) into [master](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/commit/122536d84b190f3ea7c415817b7318a99fc494a5&el=desc) will **increase** coverage by `1.18%`.
> The diff coverage is `n/a`.

[Impacted file tree graph](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/42?src=pr&el=tree)

```diff
@@            Coverage Diff             @@
##           master      #42      +/-   ##
==========================================
+ Coverage   85.89%   87.08%   +1.18%     
==========================================
  Files          34       34              
  Lines         546      635      +89     
  Branches       95      106      +11     
==========================================
+ Hits          469      553      +84     
- Misses         54       56       +2     
- Partials       23       26       +3     
```


| [Impacted Files](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/42?src=pr&el=tree) | Coverage Δ | |
|---|---|---|
| [.../src/lib/directives/show-clipped-text.directive.ts](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/42/diff?src=pr&el=tree#diff-cHJvamVjdHMvY29tcG9uZW50cy9zcmMvbGliL2RpcmVjdGl2ZXMvc2hvdy1jbGlwcGVkLXRleHQuZGlyZWN0aXZlLnRz) | `98.96% <0.00%> (ø)` | |
| [...b/directives/show-clipped-text.directive.module.ts](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/42/diff?src=pr&el=tree#diff-cHJvamVjdHMvY29tcG9uZW50cy9zcmMvbGliL2RpcmVjdGl2ZXMvc2hvdy1jbGlwcGVkLXRleHQuZGlyZWN0aXZlLm1vZHVsZS50cw==) | `100.00% <0.00%> (ø)` | |
| [...ectives/show-clipped-text.directive.test-helper.ts](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/42/diff?src=pr&el=tree#diff-cHJvamVjdHMvY29tcG9uZW50cy9zcmMvbGliL2RpcmVjdGl2ZXMvc2hvdy1jbGlwcGVkLXRleHQuZGlyZWN0aXZlLnRlc3QtaGVscGVyLnRz) | `89.13% <0.00%> (ø)` | |

------

[Continue to review full report at Codecov](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/42?src=pr&el=continue).
> **Legend** - [Click here to learn more](https://docs.codecov.io/docs/codecov-delta)
> `Δ = absolute <relative> (impact)`, `ø = not affected`, `? = missing data`
> Powered by [Codecov](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/42?src=pr&el=footer). Last update [122536d...93518af](https://codecov.io/gh/vmware/vmware-cloud-director-ui-components/pull/42?src=pr&el=lastupdated). Read the [comment docs](https://docs.codecov.io/docs/pull-request-comments).

File: projects/components/src/lib/directives/show-clipped-text.directive.ts [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+    isMouseOver: false,
+
+    /**
+     * The  that last caused the
+     */
+    get currentHost(): HTMLElement {
+        return tip.currentDirective.hostElement;
+    },
+
+    get tooltipSize(): TooltipSize {
```
Comment (ryan-bradford) [inline projects/components/src/lib/directives/show-clipped-text.directive.ts]: Fix top comment (get currentHost), and add comment to below getter (get tooltipSize)?

File: projects/components/src/lib/directives/show-clipped-text.directive.ts [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        });
+        tip.container.appendChild(tip.content);
+        document.body.appendChild(tip.container);
+        watchEvents(tip.container, tip.onMouseEnter, tip.onMouseLeave);
+        tip.content.addEventListener('transitionend', tip.onTransitionEnd);
+    },
+
+    destroy(): void {
+        document.body.removeChild(tip.container);
+        watchEvents(tip.container, tip.onMouseEnter, tip.onMouseLeave);
```
Comment (ryan-bradford) [inline projects/components/src/lib/directives/show-clipped-text.directive.ts]: Should this be `unwatchEvents`?

File: projects/components/src/lib/directives/show-clipped-text.directive.ts [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+                tip.hideTooltip(this.mouseoutDelay);
+            }
+        }
+    });
+
+    constructor(private host: ElementRef) {}
+
+    ngOnInit(): void {
+        ShowClippedTextDirective.instanceCount++;
+        tip.create();
```
Comment (ryan-bradford) [inline projects/components/src/lib/directives/show-clipped-text.directive.ts]: Can the tip singleton support multiple instances/why keep track of instanceCount?

Comment (juanmendes) [reply projects/components/src/lib/directives/show-clipped-text.directive.ts]: So we can remove it after the final directive is destroyed

File: projects/components/src/lib/directives/show-clipped-text.directive.ts (hunk 2 above)
Comment (juanmendes) [reply projects/components/src/lib/directives/show-clipped-text.directive.ts]: Definitely, will update it.

**Comment by juanmendes** [comment]
> This can be shipped but I was wondering why you had to do this `npm run build:doc-lib:scripts-dev` again in the build-app stage. And then I realized the build doc scripts are using compiled scripts from doc-lib project(and not from it's dist package) and you had to compile them again in the build app stage.
> I just want to make sure this is taken care when the doc-lib dist package is used by people outside our mono repo and i think the following will take care of that:
> `"bin": { "build-doc": "./index.js" },`
> If so, can we follow the same within our mono repo for building the docs?

After an offline conversation, we realized we can simplify it by running it directly from dist.

File: projects/components/src/lib/directives/show-clipped-text.directive.ts [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+        tip.content = document.createElement('div');
+        tip.content.classList.add('tooltip-content');
+        setStyle(tip.container, {
+            position: 'absolute',
+            zIndex: '1000',
+            opacity: '0',
+        });
+        tip.container.appendChild(tip.content);
+        document.body.appendChild(tip.container);
+        watchEvents(tip.container, tip.onMouseEnter, tip.onMouseLeave);
```
Comment (ps37) [inline projects/components/src/lib/directives/show-clipped-text.directive.ts]: Why are you watching the events here in tip.create when you are watching the events in the directive?

File: projects/components/src/lib/directives/show-clipped-text.directive.ts [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+    el.addEventListener('mouseenter', mouseIn);
+    el.addEventListener('mouseleave', mouseOut);
+}
+
+function unwatchEvents(el: HTMLElement, mouseIn: EventHandler, mouseOut: EventHandler): void {
+    el.removeEventListener('mouseenter', mouseIn);
+    el.removeEventListener('mouseleave', mouseOut);
+}
+
+function setStyle(el: HTMLElement, style: Partial<CSSStyleDeclaration>): void {
```
Comment (ps37) [inline projects/components/src/lib/directives/show-clipped-text.directive.ts]: Documentation about why this method is needed would be helpful

File: projects/components/src/lib/directives/show-clipped-text.directive.ts [hunk 6]
```diff
@@ -0,0 +1,278 @@
...
+
+    /**
+     * How long to wait before hiding the tooltip after a mouseout. This gives the user a chance to hover over the
+     * tooltip so they can copy/paste its contents
+     */
+    get mouseoutDelay(): number {
+        return tip.currentDirective.mouseoutDelay;
+    },
+
+    create(): void {
```
[the remark this answers is not in the document]
Comment (juanmendes) [reply projects/components/src/lib/directives/show-clipped-text.directive.ts]: It's private to this class, talking about where something is used is frowned upon, it should stand on its own. Do you have a question?

File: projects/components/src/lib/directives/show-clipped-text.directive.ts (hunk 5 above)
Comment (juanmendes) [reply projects/components/src/lib/directives/show-clipped-text.directive.ts]: It's clearly a small utility, private method only used here, we don't require comments for those, is something not clear?

File: projects/components/src/lib/directives/show-clipped-text.directive.ts (hunk 4 above)
Comment (juanmendes) [reply projects/components/src/lib/directives/show-clipped-text.directive.ts]: The directive watches events on the host, this watches events on the tooltip

**Pull request merged**